doc/source/dev_guide/satpy_internals.rst

```diff
@@ -0,0 +1,97 @@
+the scene object. These IDs are of type `DataID` and are imutable. They are not supposed to be used by regular users and should only be
+created in special circumstances. Satpy should take care of creating and assigning these automatically. They are also stored in the
+`attrs` of each data array as `_satpy_id`.
+
+Default and custom metadata keys
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+One thing however that the user has control over is which metadata keys are relevant to which datasets. Satpy provides two default sets
+of metadata key (or ID keys), one for regular imager bands, and the other for composites.
+The first one contains: name, wavelength, resolution, calibration, modifiers.
+The second one contains: name, resolution.
+
+As an example here is the definition of the first one in yaml:
+
+  .. code-block:: yaml
+
+    data_identification_keys:
+      name:
+        required: true
+      wavelength:
+        type: !!python/name:satpy.dataset.WavelengthRange
+      resolution:
+      calibration:
+        enum:
+            - reflectance
+            - brightness_temperature
+            - radiance
+            - counts
+      modifiers:
+        required: true
+        default: []
+        type: !!python/name:satpy.dataset.ModifierTuple
+
+To create a new set, the user can provide indications in the relevant yaml file.
+It has to be provided in header of the reader configuration file, under the `reader`
+section, as `data_identification_keys`. Each key under this is the name of relevant
+metadata key that will used to find relevant information in the attributes of the data
+arrays. Under each of this, a few options are available:
+
+ - `required`: if the item is required, False by default
+ - `type`: the type to use. More on this further down.
+ - `enum`: if the item has to be limited to a finite number of options, an enum can be used.
+   Be sure to place the options in the order of preference, with the most desirable option on top.
+ - `default`: the default value to assign to the item if nothing (or None) is provided. If this
+   option isn't provided, the key will simply be omited if it is not present in the attrs or if it
+   is None. It will be passed to the type's `convert` method if available.
+
+
+Types
+~~~~~
+Types are classes that implement a type to be used as value for metadata in the `DataID`. They have
+to implement a few methods:
+
+ - a `convert` class method that returns it's argument as an instance of the class
+ - `__hash__`, `__eq__` and `__ne__` methods
+ - a `distance` method the tells how "far" an instance of this class is from it's argument.
+
+DataID and DataQuery interactions
+=================================
+
+Different DataIDs and DataQuerys can have different metadata items defined. As such
+we define equality between different instances of these classes, and across the classes
+as equality between the sorted key/value pairs of the instances. If a DataQuery has one or more values set to `'*'`, the
```

Comment:
For equality only shared keys are taken into account, right?

Reply:
Apparently yes, and that leads to surprising behaviour:

`DataQuery(wavelength=10) == DataQuery(name="VIS006")`

Do we really want those two to compare equal?

Reply:
Indeed these should be different.

Reply:
Fixed.